### PR TITLE
GOBBLIN-1018: Report GC counts and durations from Gobblin containers …

### PR DESCRIPTION
**Pull request by sv2000**
…metrics service

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1018

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
The container health metrics service should report minor/major GC counts and durations.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Enhanced tests in ContainerHealthMetricsServiceTest class.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by sv2000** [comment]
@autumnust please review.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/ContainerHealthMetrics.java [hunk 1]
```diff
@@ -30,5 +30,10 @@
   public static final String NUM_AVAILABLE_PROCESSORS = CONTAINER_METRICS_PREFIX + "numAvailableProcessors";
   public static final String TOTAL_PHYSICAL_MEM_SIZE = CONTAINER_METRICS_PREFIX + "totalPhysicalMemSize";
   public static final String FREE_PHYSICAL_MEM_SIZE = CONTAINER_METRICS_PREFIX + "freePhysicalMemSize";
-
+  public static final String MINOR_GC_COUNT = CONTAINER_METRICS_PREFIX + "minorGcCount";
```
Comment (autumnust) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/ContainerHealthMetrics.java]: Do you need full_gc_count ( if available )

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/ContainerHealthMetricsService.java [hunk 2]
```diff
@@ -94,35 +133,75 @@ protected void runOneIteration() throws Exception {
...
     return gaugeList;
   }
 
+  private ContextAwareGauge<Double> getGauge(String name, Object metric) {
+    if (metric instanceof AtomicLong) {
+      return RootMetricContext.get().newContextAwareGauge(name, () -> Long.valueOf(((AtomicLong) metric).get()).doubleValue());
+    } else if (metric instanceof AtomicDouble) {
+      return RootMetricContext.get().newContextAwareGauge(name, () -> ((AtomicDouble) metric).get());
+    } else {
+      throw new RuntimeException(String.format("Unexpected metric type: %s for metric %s", metric.getClass().getName(), name));
```
Comment (autumnust) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/ContainerHealthMetricsService.java]: Does it need to be a vital exception ?

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/ContainerHealthMetricsService.java [hunk 3]
```diff
@@ -94,35 +133,75 @@ protected void runOneIteration() throws Exception {
...
+    gaugeList.add(getGauge(ContainerHealthMetrics.MINOR_GC_COUNT, this.minorGcCount));
+    gaugeList.add(getGauge(ContainerHealthMetrics.MINOR_GC_DURATION, this.minorGcDuration));
+    gaugeList.add(getGauge(ContainerHealthMetrics.MAJOR_GC_COUNT, this.majorGcCount));
+    gaugeList.add(getGauge(ContainerHealthMetrics.MAJOR_GC_DURATION, this.majorGcDuration));
+    gaugeList.add(getGauge(ContainerHealthMetrics.UNKNOWN_GC_COUNT, this.unknownGcCount));
+    gaugeList.add(getGauge(ContainerHealthMetrics.UNKNOWN_GC_DURATION, this.unknownGcDuration));
     return gaugeList;
   }
 
+  private ContextAwareGauge<Double> getGauge(String name, Object metric) {
```
Comment (autumnust) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/ContainerHealthMetricsService.java]: nit-pick: Is it more like a `createGauge`?

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/ContainerHealthMetricsService.java [hunk 4]
```diff
@@ -94,35 +133,75 @@ protected void runOneIteration() throws Exception {
...
+    if (metric instanceof AtomicLong) {
+      return RootMetricContext.get().newContextAwareGauge(name, () -> Long.valueOf(((AtomicLong) metric).get()).doubleValue());
```
Comment (autumnust) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/ContainerHealthMetricsService.java]: I have no strong opinion here but using `doubleValue()` for a `Long` value can save us from using `instanceof` which I believe should be avoided as much as possible in OOP.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/ContainerHealthMetricsService.java [hunk 5]
```diff
@@ -94,35 +133,75 @@ protected void runOneIteration() throws Exception {
     this.totalSwapSpaceSize.set(this.operatingSystemMXBean.getTotalSwapSpaceSize());
     this.freePhysicalMemSize.set(this.operatingSystemMXBean.getFreePhysicalMemorySize());
     this.processHeapUsedSize.set(this.memoryMXBean.getHeapMemoryUsage().getUsed());
+
+    GcStats gcStats = collectGcStats();
+    //Since GC Beans report accumulated counts/durations, we need to subtract the previous values to obtain the counts/durations
+    // since the last measurement time.
+    this.minorGcCount.set(gcStats.getMinorCount() - this.minorGcCount.get());
```
Comment (autumnust) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/ContainerHealthMetricsService.java]: Just curious, when printed, do we have any mention of this duration ?

**Comment by htran1** [comment]
+1

**Pull request closed without merging**